src/handlers.rs

```diff
@@ -1,36 +1,240 @@
 use std::ffi::CStr;
+use std::ffi::CString;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
 use url::Url;
 
-fn string_from_ptr(ptr: *const c_char) -> String {
-    let cstr = unsafe { CStr::from_ptr(ptr as *const i8) };
-    String::from(cstr.to_str().unwrap())
+// TODO(ry) Share this with the def in src/main.rs.
+#[repr(C)]
+pub struct DenoC {
+    _unused: [u8; 0],
+}
+
+// TODO(ry) Share this extern block with those in main.rs.
+// See src/reply.h
+extern "C" {
+    pub fn deno_reply_error(d: *const DenoC, cmd_id: uint32_t, msg: *const c_char);
+    pub fn deno_reply_null(d: *const DenoC, cmd_id: uint32_t);
+    pub fn deno_reply_code_fetch(
+        d: *const DenoC,
+        cmd_id: uint32_t,
+        module_name: *const c_char,
+        filename: *const c_char,
+        source_code: *const c_char,
+        output_code: *const c_char,
+    );
 }
 
+// TODO(ry) SRC_DIR is just a placeholder for future caching functionality.
+static SRC_DIR: &str = "/Users/rld/.deno/src/";
+const ASSET_PREFIX: &str = "/$asset$/";
+
 #[test]
 fn test_url() {
     let issue_list_url = Url::parse("https://github.com/rust-lang").unwrap();
     assert!(issue_list_url.scheme() == "https");
 }
 
+fn string_from_ptr(ptr: *const c_char) -> String {
+    let cstr = unsafe { CStr::from_ptr(ptr as *const i8) };
+    String::from(cstr.to_str().unwrap())
+}
+
+fn as_cstring(s: &String) -> CString {
+    CString::new(s.as_str()).unwrap()
+}
+
+// Prototype: https://github.com/ry/deno/blob/golang/os.go#L56-L68
+#[allow(dead_code)]
+fn src_file_to_url<P: AsRef<Path>>(filename: P) -> String {
+    assert!(SRC_DIR.len() > 0, "SRC_DIR shouldn't be empty");
+
+    let filename = filename.as_ref().to_path_buf();
+    let src = (SRC_DIR.as_ref() as &Path).to_path_buf();
+
+    if filename.starts_with(&src) {
+        let rest = filename.strip_prefix(&src).unwrap();
+        "http://".to_string() + rest.to_str().unwrap()
+    } else {
+        String::from(filename.to_str().unwrap())
+    }
+}
+
+#[test]
+fn test_src_file_to_url() {
+    assert_eq!("hello", src_file_to_url("hello"));
+    assert_eq!("/hello", src_file_to_url("/hello"));
+    let x = SRC_DIR.to_string() + "hello";
+    assert_eq!("http://hello", src_file_to_url(&x));
+    let x = SRC_DIR.to_string() + "/hello";
+    assert_eq!("http://hello", src_file_to_url(&x));
+}
+
+// Prototype: https://github.com/ry/deno/blob/golang/os.go#L70-L98
+// Returns (module name, local filename)
+fn resolve_module(
+    module_specifier: &String,
+    containing_file: &String,
+) -> Result<(String, String), url::ParseError> {
+    info!(
+        "resolve_module before module_specifier {} containing_file {}",
+        module_specifier, containing_file
+    );
+
+    //let module_specifier = src_file_to_url(module_specifier);
+    //let containing_file = src_file_to_url(containing_file);
+    //let base_url = Url::parse(&containing_file)?;
+
+    let j: Url = if containing_file.as_str().ends_with("/") {
```

Comment:
* Is containing_file a filename or a url?
* I don't understand how containing_file could end with a slash.

Reply:
It's a path or URL depending on where the resource is being queried from. One place where the trailing slash is used is in main.ts:
```
const mod = runtime.resolveModule(inputFn, `${cwd}/`);
```
I'm not trying to improve the semantics or logic of these functions at this point - the aim is to d a direct port of the Go code - later we can improve how it actually works. (I think the exact meaning of `containing_file` should indeed be better clarified in the future.)